core/src/main/java/org/apache/iceberg/util/FileSystemWalker.java

```diff
@@ -0,0 +1,216 @@
+import org.apache.iceberg.io.FileInfo;
+import org.apache.iceberg.io.SupportsPrefixOperations;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+
+/**
+ * Utility class for recursively traversing file systems and identifying hidden paths. Provides
+ * methods to list files recursively while filtering out hidden paths based on specified criteria.
+ */
+public class FileSystemWalker {
+
+  private FileSystemWalker() {}
+
+  /**
+   * Recursively lists files in the specified directory that satisfy the given conditions. Use
+   * {@link PartitionAwareHiddenPathFilter} to filter out hidden paths.
+   *
+   * @param io File system interface supporting prefix operations
+   * @param dir Base directory to start recursive listing
+   * @param specs Map of {@link PartitionSpec partition specs} for this table.
+   * @param predicate Additional filter condition for files
+   * @param consumer Consumer to accept matching file locations
+   */
+  public static void listDirRecursivelyWithFileIO(
+      SupportsPrefixOperations io,
+      String dir,
+      Map<Integer, PartitionSpec> specs,
+      Predicate<FileInfo> predicate,
+      Consumer<String> consumer) {
+    PathFilter pathFilter = PartitionAwareHiddenPathFilter.forSpecs(specs);
+    String listPath = dir;
+    if (!dir.endsWith("/")) {
+      listPath = dir + "/";
+    }
+
+    Iterable<FileInfo> files = io.listPrefix(listPath);
+    for (FileInfo file : files) {
+      Path path = new Path(file.location());
+      if (!isHiddenPath(dir, path, pathFilter) && predicate.test(file)) {
+        consumer.accept(file.location());
+      }
+    }
+  }
+
+  /**
+   * Recursively traverses the specified directory using Hadoop API to collect file paths that meet
+   * the conditions.
+   *
+   * <p>This method provides depth control and subdirectory quantity limitation:
+   *
+   * <ul>
+   *   <li>Stops traversal when maximum recursion depth is reached and adds current directory to
+   *       pending list
+   *   <li>Stops traversal when number of direct subdirectories exceeds threshold and adds
+   *       subdirectories to pending list
+   * </ul>
+   *
+   * @param dir The starting directory path to traverse
+   * @param specs Map of {@link PartitionSpec partition specs} for this table.
+   * @param predicate File filter condition, only files satisfying this condition will be collected
+   * @param conf Hadoop conf
+   * @param maxDepth Maximum recursion depth limit
+   * @param maxDirectSubDirs Upper limit of subdirectories that can be processed directly
+   * @param remainingSubDirs Output parameter for storing unprocessed directory paths
+   * @param consumer Consumer for collecting qualified file paths
+   */
+  public static void listDirRecursivelyWithHadoop(
+      String dir,
+      Map<Integer, PartitionSpec> specs,
+      Predicate<FileStatus> predicate,
+      Configuration conf,
+      int maxDepth,
+      int maxDirectSubDirs,
+      List<String> remainingSubDirs,
```

Comment:
Just an idea, not sure it is a good one, could this be:
```
Consumer<String> remainingSubDirs,
Consumer<String> files) {
```

Maybe this is a bit more consistent API